### ISSUE TITLE
Discuss logging (default configuration on Linux, use of systemd-journald)

### ISSUE THREAD

**Issue by julianbrost** [Icinga/icinga2#8857]
With 2.13, we will change the default logging configuration on Windows to use the Windows Event Log (#8710). The purpose of this issue is to discuss how to move on with logging on other platforms.

Some ideas:
* Switch to syslog by default, there shouldn't really be a need for applications to write their own log files. Given that systemd-journald is now quite widespread on Linux, it also got way nicer to interact with these logs.
* We could think of implementing a `JournaldLogger` (or integrate this into the `SyslogLogger`) making use of the [sd-journal](https://www.freedesktop.org/software/systemd/man/sd-journal.html) API. This would for example allow adding more metadata to log entries without having to pollute the log message itself (you can see this additional information for example in `journalctl -o json-pretty`).

**Comment by widhalmt** [comment]
I like the idea of having more meta data, especially when you still could keep the existing messages intact.

One of the things, I hear a lot from users but still have no cool idea how to implement it, is to have a way more detailed selection of what to log and what not to log. See, some setups are so big, they can't turn on `debug` mode without breaking their storage systems but they need *some* data from there. e.g. one customer wanted to explicitly log every interaction with the API but on the other hand still keep the log on `warning` level. 

From a logmanagement point of view, the most important thing about this idea is that it would be very helpful to have a way of getting the same strings out of different formats. What I mean is, that logmanagement usually works like a Unix pipe where you can chain filters and transformations as well as reuse parsing code. So, when I take the new log events, I should be able to strip (and maybe save elsewhere) the metadata and end up with the same string where the same rules can apply. Be it via JournalD, Event Log, JSON objects in a file or single lines in syslog format.

**Comment by haxtibal** [comment]
> We could think of implementing a JournaldLogger [...] making use of the sd-journal API

Very much liked the idea and started to implement exactly that: https://github.com/haxtibal/icinga2/tree/feature/journaldlogger Will send a PR when it's ready. Shouldn't take too long, it's a pleasure to work with your code base.

We can then store structured and indexed meta data on almost any modern Linux system, and e.g. watch the `ApiListener` with `journalctl -f ICINGA2_FACILITY=ApiListener`, while we get syslog compatibility for free.

**Comment by Al2Klimov** [comment]
> there shouldn't really be a need for applications to write their own log files

Tbh I'm a bit sceptical regarding that statement as-is. Actually a lot of apps write own logs, from HTTPd to MySQL.

@bobapple Your opinion, please.

**Comment by julianbrost** [comment]
Well, that statement maybe is a bit extreme and heavily opinionated, so let me elaborate a bit.

First of all, I meant to refer to general application logging like "this error happened" and so on. Sure, there are more specialized logs where logging them to a separate file makes sense, for example web server access logs. Like there might be some tooling that expects the files in this very specific format, these are likely to contain personally identifiable information so you might want to keep them separate, etc.

But for the average log messages, the system provides facilities for logging, so I think applications should just use them. I want to be able to `systemctl status something` if a service didn't start and see its log output. I want to be able to do `journalctl -f -p err` and see an error scroll by instead of having to check multiple error log files (for example application, php-fpm, web server).

The main issue I see with this suggestion in the context of icinga2 is that the logging is quite verbose. For example, if some agent is down and a node can't connect to it, maybe writing an error message about that to syslog/journald every 10 seconds is a bit excessive (but so is writing it to `icinga2.log`, there it just tends to annoy people less).

**Comment by Al2Klimov** [comment]
OK, shall we do pretty much the same as for Windows?

* Linux: JournaldLogger
* \*nix: syslog

**Comment by julianbrost** [comment]
Given that we received feedback after switching to the Windows Event Log by default that this were too many messages (for example #8996, but I remember there were also customer support tickets), I don't think we're at a point where doing this on Linux would be a good idea. On Windows, it's only an agent, a satellite or master will generate quite a lot more messages than an agent.

Using syslog/journald is something I'd like to see in the long run, but just switching the default without improving the quality of log messages just won't do it I fear.

**Comment by julianbrost** [comment]
Just wrote https://github.com/Icinga/icinga2/pull/9597#issuecomment-1330709159 which is also relevant here.

tl;dr: I don't think switching the default logging mechanism in 2.14 is something we can realistically do without causing frustration. Instead, we should focus on improving the quality of the logs first while keeping the log file as default. Once this has proven in practice to be suitable for also logging into syslog/journald we should consider switching the default.
